### PR TITLE
Upgrade to Lucene 9.12.0

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

**Review by romseygeek — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
@romseygeek Could you please take another look? I've made some changes required due to the new `READONCE` restriction which has been introduced. (used https://github.com/opensearch-project/OpenSearch/pull/15333 as guide).

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -982,13 +983,19 @@ void sendFiles(Store store, StoreFileMetadata[] files, IntSupplier translogOps,
...
+                    // https://github.com/apache/lucene/blob/b2d3a2b37e00f19a74949097736be8fd64745f61/lucene/test-framework/src/java/org/apache/lucene/tests/store/MockDirectoryWrapper.java#L817
+                    if (md.name().startsWith(IndexFileNames.SEGMENTS) == false) {
+                        final IndexInput indexInput = store.directory().openInput(md.name(), IOContext.READ);
+                        currentInput = new InputStreamIndexInput(indexInput, md.length()) {
+                            @Override
+                            public void close() throws IOException {
+                                IOUtils.close(indexInput, super::close); // InputStreamIndexInput's close is a noop
+                            }
+                        };
+                    }
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:998]: Does this need an `else` clause?  As I read it this means that we don't set `currentInput` at all when the new resource is a segments file?

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:998]: Oh hang on, I see that this is done in `readBytes()`.  It feels a bit awkward to split it across two methods though?

Comment (matriv) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:998]: I've tried this: https://github.com/crate/crate/pull/16805 but it leads to multiple failures.

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:998]: Thanks, probably not worth spending any more time on then.

Comment (matriv) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:998]: thx!

**Pull request merged**